include/nbl/video/IGPUCommandBuffer.h

```diff
@@ -227,6 +232,93 @@ class IGPUCommandBuffer :
             for (uint32_t i = _first + 1u; i < IGPUPipelineLayout::DESCRIPTOR_SET_COUNT; ++i)
                 _destPplnLayouts[i] = nullptr;
     }
+
+    template <typename Cmd, typename... Args>
+    Cmd* emplace(Args&&... args)
+    {
+        if (m_segmentListTail == nullptr)
+        {
+            void* cmdSegmentMem = m_cmdpool->m_commandSegmentPool.allocate(IGPUCommandPool::COMMAND_SEGMENT_SIZE, alignof(IGPUCommandPool::CommandSegment));
+            if (!cmdSegmentMem)
+                return nullptr;
+
+            m_segmentListTail = new (cmdSegmentMem) IGPUCommandPool::CommandSegment;
+            m_segmentListHeadItr.m_segment = m_segmentListTail;
+        }
+
+        // Cmd* cmd = m_segmentListTail->allocate<Cmd, Args...>(args...);
+        Cmd* cmd = m_segmentListTail->allocate<Cmd>(std::forward<Args>(args)...);
+        if (!cmd)
+        {
+            void* nextSegmentMem = m_cmdpool->m_commandSegmentPool.allocate(IGPUCommandPool::COMMAND_SEGMENT_SIZE, alignof(IGPUCommandPool::CommandSegment));
+            if (nextSegmentMem == nullptr)
+                return nullptr;
+
+            IGPUCommandPool::CommandSegment* nextSegment = new (nextSegmentMem) IGPUCommandPool::CommandSegment;
+
+            // cmd = m_segmentListTail->allocate<Cmd, Args...>(args...);
+            cmd = m_segmentListTail->allocate<Cmd>(std::forward<Args>(args)...);
```

Comment:
you can't forward the same args twice, this is why my original code snippet split allocation (because allocation would only take const references) from construction/placement new (because it would only be attempted once, so args would be moved once if r-values)

Reply:
> you can't forward the same args twice

Ah nice, I didn't know that.
Though I find it a bit strange that the compiler can't deduce that only of the codepaths containing `std::forward` will ever run.

Anyway, this and this comment https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r863838228 makes perfect sense, thank you.

Reply:
Resolved.